## Week 4, Lab 1: Predicting Left-Handedness from Psychological Factors
> Author: Matt Brems

We can sketch out the data science process as follows:
1. Define the problem.
2. Obtain the data.
3. Explore the data.
4. Model the data.
5. Evaluate the model.
6. Answer the problem.

We'll walk through a full data science problem in this lab. 

---
## Step 1: Define The Problem.

You're currently a data scientist working at a university. A professor of psychology is attempting to study the relationship between personalities and left-handedness. They have tasked you with gathering evidence so that they may publish.

Specifically, the professor says "I need to prove that left-handedness is caused by some personality trait. Go find that personality trait and the data to back it up."

As a data scientist, you know that any real data science problem must be **specific** and **conclusively answerable**. For example:
- Bad data science problem: "What is the link between obesity and blood pressure?"
    - This is vague and is not conclusively answerable. That is, two people might look at the conclusion and one may say "Sure, the problem has been answered!" and the other may say "The problem has not yet been answered."
- Good data science problem: "Does an association exist between obesity and blood pressure?"
    - This is more specific and is conclusively answerable. The problem specifically is asking for a "Yes" or "No" answer. Based on that, two independent people should both be able to say either "Yes, the problem has been answered" or "No, the problem has not yet been answered."
- Excellent data science problem: "As obesity increases, how does blood pressure change?"
    - This is very specific and is conclusively answerable. The problem specifically seeks to understand the effect of one variable on the other.

### 1. In the context of the left-handedness and personality example, what are three specific and conclusively answerable problems that you could answer using data science? 

> You might find it helpful to check out the codebook in the repo for some inspiration.

### Answer:

* 1) Whether someone is more likely to be a left_handedness if he/she hates shopping more? (Q33)
* 2) Whether someone is more likely to be a left_handedness if he/she likes dancing more? (Q40)
* 3) Whether someone is more likely to be a left_handedness if he/she likes guns more? (Q9)


---
## Step 2: Obtain the data.

### 2. Read in the file titled "data.csv."
> Hint: Despite being saved as a .csv file, you won't be able to simply `pd.read_csv()` this data!

In [97]:
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split

In [82]:
df = pd.read_csv('data.csv',delimiter='\t')

### 3. Suppose that, instead of us giving you this data in a file, you were actually conducting a survey to gather this data yourself. From an ethics/privacy point of view, what are three things you might consider when attempting to gather this data?
> When working with sensitive data like sexual orientation or gender identity, we need to consider how this data could be used if it fell into the wrong hands!

Answer:
* 1) How to store the data in order to make it safe and not including personal info like names.
* 2) Make sure the samle is random
* 3) Create a confortable environment for anyone who answear my survey so he/she can give the real answers.

---
## Step 3: Explore the data.

### 4. Conduct exploratory data analysis on this dataset.
> If you haven't already, be sure to check out the codebook in the repo, as that will help in your EDA process.

In [67]:
df.describe().T

# There are etrem values in age,introelapse and testelapse. However, no missing values.

count        mean          std   min    25%    50%     75%  \
Q1           4184.0    1.962715     1.360291   0.0    1.0    1.0    3.00   
Q2           4184.0    3.829589     1.551683   0.0    3.0    5.0    5.00   
Q3           4184.0    2.846558     1.664804   0.0    1.0    3.0    5.00   
Q4           4184.0    3.186902     1.476879   0.0    2.0    3.0    5.00   
Q5           4184.0    2.865440     1.545798   0.0    1.0    3.0    4.00   
Q6           4184.0    3.672084     1.342238   0.0    3.0    4.0    5.00   
Q7           4184.0    3.216539     1.490733   0.0    2.0    3.0    5.00   
Q8           4184.0    3.184512     1.387382   0.0    2.0    3.0    4.00   
Q9           4184.0    2.761233     1.511805   0.0    1.0    3.0    4.00   
Q10          4184.0    3.522945     1.242890   0.0    3.0    4.0    5.00   
Q11          4184.0    2.748805     1.443078   0.0    1.0    3.0    4.00   
Q12          4184.0    2.852772     1.556284   0.0    1.0    3.0    4.00   
Q13          4184.0    2.657505     1.559575   0.0    1.0    3.0    4.00   
Q14          4184.0    3.334130     1.522866   0.0    2.0    4.0    5.00   
Q15          4184.0    3.168021     1.501683   0.0    2.0    3.0    5.00   
Q16          4184.0    2.930210     1.575544   0.0    1.0    3.0    4.00   
Q17          4184.0    2.564771     1.619010   0.0    1.0    2.0    4.00   
Q18          4184.0    3.424952     1.413236   0.0    2.0    4.0    5.00   
Q19          4184.0    2.928537     1.493122   0.0    2.0    3.0    4.00   
Q20          4184.0    3.639818     1.414569   0.0    3.0    4.0    5.00   
Q21          4184.0    2.867591     1.360858   0.0    2.0    3.0    4.00   
Q22          4184.0    3.595124     1.354475   0.0    3.0    4.0    5.00   
Q23          4184.0    3.861138     1.291425   0.0    3.0    4.0    5.00   
Q24          4184.0    3.337237     1.426095   0.0    2.0    3.0    5.00   
Q25          4184.0    1.999761     1.290747   0.0    1.0    1.0    3.00   
Q26          4184.0    3.001434     1.480610   0.0    2.0    3.0    4.00   
Q27          4184.0    2.730641     1.485883   0.0    1.0    3.0    4.00   
Q28          4184.0    2.624044     1.481709   0.0    1.0    3.0    4.00   
Q29          4184.0    2.543738     1.611428   0.0    1.0    2.0    4.00   
Q30          4184.0    2.894359     1.477968   0.0    1.0    3.0    4.00   
Q31          4184.0    3.002151     1.420032   0.0    2.0    3.0    4.00   
Q32          4184.0    2.869503     1.659141   0.0    1.0    3.0    5.00   
Q33          4184.0    2.741874     1.405670   0.0    1.0    3.0    4.00   
Q34          4184.0    3.022228     1.562694   0.0    1.0    3.0    4.00   
Q35          4184.0    3.074092     1.546400   0.0    1.0    3.0    5.00   
Q36          4184.0    2.610660     1.409707   0.0    1.0    2.0    4.00   
Q37          4184.0    3.465344     1.521460   0.0    2.0    4.0    5.00   
Q38          4184.0    2.798757     1.413584   0.0    1.0    3.0    4.00   
Q39          4184.0    2.569312     1.621772   0.0    1.0    2.0    4.00   
Q40          4184.0    2.984226     1.483752   0.0    2.0    3.0    4.00   
Q41          4184.0    3.385277     1.423055   0.0    2.0    4.0    5.00   
Q42          4184.0    2.704828     1.544345   0.0    1.0    3.0    4.00   
Q43          4184.0    2.676386     1.523097   0.0    1.0    3.0    4.00   
Q44          4184.0    2.736616     1.471845   0.0    1.0    3.0    4.00   
introelapse  4184.0  347.808556  5908.901681   1.0    6.0   12.0   35.00   
testelapse   4184.0  479.994503  3142.178542   7.0  186.0  242.0  324.25   
fromgoogle   4184.0    1.576243     0.494212   1.0    1.0    2.0    2.00   
engnat       4184.0    1.239962     0.440882   0.0    1.0    1.0    1.00   
age          4184.0   30.370698   367.201726  13.0   18.0   21.0   27.00   
education    4184.0    2.317878     0.874264   0.0    2.0    2.0    3.00   
gender       4184.0    1.654398     0.640915   0.0    1.0    2.0    2.00   
orientation  4184.0    1.833413     1.303454   0.0    1.0    1

In [83]:
# remove country column and fromgoogle
df.drop(['country','fromgoogle'],axis=1,inplace=True)

In [84]:
df['introelapse'].describe()

count      4184.000000
mean        347.808556
std        5908.901681
min           1.000000
25%           6.000000
50%          12.000000
75%          35.000000
max      252063.000000
Name: introelapse, dtype: float64

In [85]:
df['testelapse'].describe()

count      4184.000000
mean        479.994503
std        3142.178542
min           7.000000
25%         186.000000
50%         242.000000
75%         324.250000
max      119834.000000
Name: testelapse, dtype: float64

In [86]:
# find the outliers of introelapse and reove it from the obvervations
introelapse_iqr = 35 - 6
introelapse_outliers = 35 + 1.5 * introelapse_iqr
df = df[df['introelapse'] <= introelapse_outliers]

In [87]:
# find the outliers of testelapse and reove it from the obvervations, this time we also need to remove ouliers which have extremly low value,
# because if someone answers the question extremly quick, we would consider he didn't pay too much attention on it.
testelapse_iqr = 319 - 185

testelapse_outliers_upper = 319 + 1.5 * testelapse_iqr
testelapse_outliers_bottom = 185 - 1.5 * testelapse_iqr
df = df[(df['testelapse'] <= testelapse_outliers_upper) & (df['testelapse'] >= testelapse_outliers_bottom)]

In [88]:
# remove observations that age is outliers
df = df[df['age'] < 200]

In [89]:
X = df.drop('hand',axis=1)
y = df['hand'].map(lambda x:1 if x ==2 else 0)

In [90]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X = sc.fit_transform(X)

In [98]:
X_train,X_test,y_train,y_test = train_test_split(X,y,random_state=42)

---
## Step 4: Model the data.

### 5. Suppose I wanted to use Q1 - Q44 to predict whether or not the person is left-handed. Would this be a classification or regression problem? Why?

Answer: Classification

### 6. We want to use $k$-nearest neighbors to predict whether or not a person is left-handed based on their responses to Q1 - Q44. Before doing that, however, you remember that it is often a good idea to standardize your variables. In general, why would we standardize our variables? Give an example of when we would standardize our variables.

Answer: Standarlize data will make the magnitudes of each predictors in a reasonable range. For instance, if we want to use age and income to predict someone's education level. Age must be in the range of [0,200] (hopefully) but someone's income could be billions of dolloar. It would be hard to estimate the coefs.(Inmagine the extremly long contours)

### 7. Give an example of when we might not standardize our variables.

Answer: The answers for Q1- Q44 is a good example since every answer is in the range of 0 to 5 so there is no magnititude problem.

### 8. Based on your answers to 6 and 7, do you think we should standardize our predictor variables in this case? Why or why not?

Answer: We need to standarlize some of the predictors, such as introelapse,testelapse.

### 9. We want to use $k$-nearest neighbors to predict whether or not a person is left-handed. What munging/cleaning do we need to do to our $y$ variable in order to explicitly answer this question? Do it.

Answer: Encode the y(hand) into binary: 1 = left-handed, 0 = not left-handed

### 10. The professor for whom you work suggests that you set $k = 4$. In this specific case, why might this be a bad idea?

Answer: There are 4184 observations, choose the k = sqrt(4000) is a good point to start.

### 11. Let's *(finally)* use $k$-nearest neighbors to predict whether or not a person is left-handed!

> Be sure to create a train/test split with your data!

> Create four separate models, one with $k = 3$, one with $k = 5$, one with $k = 15$, and one with $k = 25$.

> Instantiate and fit your models.

In [76]:
from sklearn.neighbors import KNeighborsClassifier

In [78]:
knn_3 = KNeighborsClassifier(n_neighbors=3)
knn_5 = KNeighborsClassifier(n_neighbors=5)
knn_15 = KNeighborsClassifier(n_neighbors=15)
knn_25 = KNeighborsClassifier(n_neighbors=25)

In [99]:
knn_3.fit(X_train,y_train)
knn_5.fit(X_train,y_train)
knn_15.fit(X_train,y_train)
knn_25.fit(X_train,y_train)

KNeighborsClassifier(n_neighbors=25)

Being good data scientists, we know that we might not run just one type of model. We might run many different models and see which is best.

### 12. We want to use logistic regression to predict whether or not a person is left-handed. Before we do that, let's check the [documentation for logistic regression in sklearn](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html). Is there default regularization? If so, what is it? If not, how do you know?

Answer: Yes, there is a defualt regularization. The default regularization is L2: Ridge. I know because in the documentation, the first parameter of LogisticRegression is penalty which is set to l2 as default value.

### 13. We want to use logistic regression to predict whether or not a person is left-handed. Before we do that, should we standardize our features?

Answer:Yes, we do need to standarlize our features.

### 14. Let's use logistic regression to predict whether or not the person is left-handed.


> Be sure to use the same train/test split with your data as with your $k$-NN model above!

> Create four separate models, one with LASSO and $\alpha = 1$, one with LASSO and $\alpha = 10$, one with Ridge and $\alpha = 1$, and one with Ridge and $\alpha = 10$. *(Hint: Be careful with how you specify $\alpha$ in your model!)*

> Instantiate and fit your models.

In [92]:
from sklearn.linear_model import LogisticRegression

In [95]:
logreg_1 = LogisticRegression(penalty='l1',C = 1,solver='liblinear')
logreg_2 = LogisticRegression(penalty='l1',C = 10,solver='liblinear')
logreg_3 = LogisticRegression(penalty='l2',C = 1)
logreg_4 = LogisticRegression(penalty='l2',C = 10)

In [100]:
logreg_1.fit(X_train,y_train)
logreg_2.fit(X_train,y_train)
logreg_3.fit(X_train,y_train)
logreg_4.fit(X_train,y_train)

LogisticRegression(C=10)

---
## Step 5: Evaluate the model(s).

### 15. Before calculating any score on your data, take a step back. Think about your $X$ variable and your $Y$ variable. Do you think your $X$ variables will do a good job of predicting your $Y$ variable? Why or why not? What impact do you think this will have on your scores?

Answer: I don't think it will do a good job since I feel that X's are not correlated with Y

### 16. Using accuracy as your metric, evaluate all eight of your models on both the training and testing sets. Put your scores below. (If you want to be fancy and generate a table in Markdown, there's a [Markdown table generator site linked here](https://www.tablesgenerator.com/markdown_tables#).)
- Note: Your answers here might look a little weird. You didn't do anything wrong; that's to be expected!

In [102]:
print(knn_3.score(X_train,y_train),knn_5.score(X_train,y_train),knn_15.score(X_train,y_train),knn_25.score(X_train,y_train))

print(logreg_1.score(X_train,y_train), logreg_2.score(X_train,y_train), logreg_3.score(X_train,y_train), logreg_4.score(X_train,y_train))

0.9102105681366707 0.8947159316646802 0.8883591577274533 0.8883591577274533
0.8883591577274533 0.8883591577274533 0.8883591577274533 0.8883591577274533


In [101]:
print(knn_3.score(X_test,y_test),knn_5.score(X_test,y_test),knn_15.score(X_test,y_test),knn_25.score(X_test,y_test))

print(logreg_1.score(X_test,y_test), logreg_2.score(X_test,y_test), logreg_3.score(X_test,y_test), logreg_4.score(X_test,y_test))

0.8569725864123957 0.8688915375446961 0.8808104886769964 0.8808104886769964
0.8808104886769964 0.8808104886769964 0.8808104886769964 0.8808104886769964


### 17. In which of your $k$-NN models is there evidence of overfitting? How do you know?

Answer: When k=3 or k=5, the models are overfitting, because train test score are higher than test set score.

### 18. Broadly speaking, how does the value of $k$ in $k$-NN affect the bias-variance tradeoff? (i.e. As $k$ increases, how are bias and variance affected?)

Answer: When K increase, the bias incrase but variance decrease.

### 19. If you have a $k$-NN model that has evidence of overfitting, what are three things you might try to do to combat overfitting?

Answer:
* 1) K is too small
* 2) Get more obsevations
* 3) Find out nosises and remove them

### 20. In which of your logistic regression models is there evidence of overfitting? How do you know?

Answer: No evidence of overfitting.

### 21. Broadly speaking, how does the value of $C$ in logistic regression affect the bias-variance tradeoff? (i.e. As $C$ increases, how are bias and variance affected?)

Answer: The C doesn't do anything at all.

### 22. For your logistic regression models, play around with the regularization hyperparameter, $C$. As you vary $C$, what happens to the fit and coefficients in the model? What do you think this means in the context of this specific problem?

Answer: Since the logistic regression models are not even better than the null model. So the change of C has no affections on the model.

### 23. If you have a logistic regression model that has evidence of overfitting, what are three things you might try to do to combat overfitting?

Answer:
* 1) Tuning the hyperparameters
* 2) remove non-related features
* 3) collect more data

---
## Step 6: Answer the problem.

### 24. Suppose you want to understand which psychological features are most important in determining left-handedness. Would you rather use $k$-NN or logistic regression? Why?

Answer: Loistic Regression, because the coefs in logistic regression can be interpreted. The larger the absolute value of coef of a predictor, the more importeant the feature is.

### 25. Select your logistic regression model that utilized LASSO regularization with $\alpha = 1$. Interpret the coefficient for `Q1`.

In [112]:
pd.DataFrame([df.drop('hand',axis=1).columns,logreg_1.coef_.reshape(-1,1)]).T.loc[0]

0                        Q1
1    [-0.09784574576524913]
Name: 0, dtype: object

Answer: 1 unit increases in Q1 will result the log(odds) of being left_handed decrease -0.0978

### 26. If you have to select one model overall to be your *best* model, which model would you select? Why?
- Usually in the "real world," you'll fit many types of models but ultimately need to pick only one! (For example, a client may not understand what it means to have multiple models, or if you're using an algorithm to make a decision, it's probably pretty challenging to use two or more algorithms simultaneously.) It's not always an easy choice, but you'll have to make it soon enough. Pick a model and defend why you picked this model!

Answer: Logistic regression. Even thought none of the models are doing very well, logistics regression can be easily interpreted.

### 27. Circle back to the three specific and conclusively answerable questions you came up with in Q1. Answer one of these for the professor based on the model you selected!

Answer: None of the features collected by the dataset is related to left handeness. I guess it's more like a genetic problem.

### BONUS:
Looking for more to do? Probably not - you're busy! But if you want to, consider exploring the following. (They could make for a blog post!)
- Create a visual plot comparing training and test metrics for various values of $k$ and various regularization schemes in logistic regression.
- Rather than just evaluating models based on accuracy, consider using sensitivity, specificity, etc.
- In the context of predicting left-handedness, why are unbalanced classes concerning? If you were to re-do this process given those concerns, what changes might you make?
- Fit and evaluate a generalized linear model other than logistic regression (e.g. Poisson regression).
- Suppose this data were in a `SQL` database named `data` and a table named `inventory`. What `SQL` query would return the count of people who were right-handed, left-handed, both, or missing with their class labels of 1, 2, 3, and 0, respectively? (You can assume you've already logged into the database.)